libImaging/Resample.c

```diff
@@ -65,125 +50,128 @@ static inline float bicubic_filter(float x)
     if (filterscale < 1.0) {
         filterscale = 1.0;
     }
 
-    support = support * filterscale;
+    /* determine support size (length of resampling filter) */
+    support = filterp->support * filterscale;
 
     /* maximum number of coofs */
     kmax = (int) ceil(support) * 2 + 1;
 
     // check for overflow
-    if (kmax > 0 && xsize > SIZE_MAX / kmax)
-        return (Imaging) ImagingError_MemoryError();
+    if (outSize > SIZE_MAX / (kmax * sizeof(double)))
+        return 0;
 
-    // sizeof(float) should be greater than 0
-    if (xsize * kmax > SIZE_MAX / sizeof(float))
-        return (Imaging) ImagingError_MemoryError();
+    // sizeof(double) should be greater than 0 as well
+    if (outSize > SIZE_MAX / (2 * sizeof(double)))
+        return 0;
 
     /* coefficient buffer */
-    kk = malloc(xsize * kmax * sizeof(float));
+    kk = malloc(outSize * kmax * sizeof(double));
     if ( ! kk)
-        return (Imaging) ImagingError_MemoryError();
-
-    // sizeof(int) should be greater than 0 as well
-    if (xsize > SIZE_MAX / (2 * sizeof(int)))
-        return (Imaging) ImagingError_MemoryError();
+        return 0;
 
-    xbounds = malloc(xsize * 2 * sizeof(int));
+    xbounds = malloc(outSize * 2 * sizeof(int));
     if ( ! xbounds) {
         free(kk);
-        return (Imaging) ImagingError_MemoryError();
+        return 0;
     }
 
-    for (xx = 0; xx < xsize; xx++) {
-        k = &kk[xx * kmax];
+    for (xx = 0; xx < outSize; xx++) {
         center = (xx + 0.5) * scale;
         ww = 0.0;
         ss = 1.0 / filterscale;
         xmin = (int) floor(center - support);
         if (xmin < 0)
             xmin = 0;
         xmax = (int) ceil(center + support);
-        if (xmax > imIn->xsize)
-            xmax = imIn->xsize;
-        for (x = xmin; x < xmax; x++) {
-            float w = filterp->filter((x - center + 0.5) * ss) * ss;
-            k[x - xmin] = w;
+        if (xmax > inSize)
+            xmax = inSize;
+        xmax -= xmin;
+        k = &kk[xx * kmax];
+        for (x = 0; x < xmax; x++) {
+            double w = filterp->filter((x + xmin - center + 0.5) * ss);
+            k[x] = w;
             ww += w;
         }
-        for (x = 0; x < xmax - xmin; x++) {
+        for (x = 0; x < xmax; x++) {
             if (ww != 0.0)
                 k[x] /= ww;
         }
         xbounds[xx * 2 + 0] = xmin;
         xbounds[xx * 2 + 1] = xmax;
     }
+    *xboundsp = xbounds;
+    *kkp = kk;
+    return kmax;
+}
+
+
+Imaging
+ImagingResampleHorizontal_8bpc(Imaging imIn, int xsize, struct filter *filterp)
+{
+    ImagingSectionCookie cookie;
+    Imaging imOut;
+    int ss0, ss1, ss2, ss3;
+    int xx, yy, x, kmax, xmin, xmax;
+    int *xbounds;
+    int *k, *kk;
+    double *prekk;
+
+
+    kmax = ImagingPrecompute(imIn->xsize, xsize, filterp, &xbounds, &prekk);
+    if ( ! kmax) {
+        return (Imaging) ImagingError_MemoryError();
+    }
+    
+    kk = malloc(xsize * kmax * sizeof(int));
```

Comment:
Need to check this against INT_MAX for overflow.  Also probably convert to calloc.